gms/gossiper.cc

```diff
@@ -582,49 +582,84 @@ void gossiper::notify_failure_detector(inet_address endpoint, const endpoint_sta
 }
 
 // Runs inside seastar::async context
-void gossiper::apply_state_locally_without_listener_notification(std::unordered_map<inet_address, endpoint_state> map) {
-    for (auto& x : map) {
-        const inet_address& node = x.first;
-        const endpoint_state& remote_state = x.second;
-        if (node == this->get_broadcast_address()) {
-            continue;
-        }
-        auto permit = this->lock_endpoint(node).get0();
-        auto es = this->get_endpoint_state_for_endpoint_ptr(node);
-        if (es) {
-            endpoint_state& local_state = *es;
-            int local_generation = local_state.get_heart_beat_state().get_generation();
-            int remote_generation = remote_state.get_heart_beat_state().get_generation();
-            logger.debug("{} local generation {}, remote generation {}", node, local_generation, remote_generation);
-            if (remote_generation > utils::get_generation_number() + MAX_GENERATION_DIFFERENCE) {
-                logger.warn("received an invalid gossip generation for peer {}; local generation = {}, received generation = {}",
-                    node, local_generation, remote_generation);
-            } else if (remote_generation > local_generation) {
-                logger.info("Applying remote_state for node {} (remote generation > local generation)", node);
+void gossiper::do_apply_state_locally(gms::inet_address node, const endpoint_state& remote_state, bool listener_notification) {
+    // If state does not exist just add it. If it does then add it if the remote generation is greater.
+    // If there is a generation tie, attempt to break it by heartbeat version.
```

Comment:
It's impossible to review this large code chunks. Can you just change apply_state_locally() directly (perhaps renaming it to do_apply_state_locally with an extra parameter).

Reply:
do_apply_state_locally extracts the common code. It is much easier to review without touching the existing apply_state_locally. But you suggested to merge them which unavoidably makes the patch set larger to review. 

The code t is not that complicated: it is like compare generation hen compare version.

Reply:
For me it's harder. There are many lines moved, and I can't detect what the incremental changes are.

Reply:
The code inside listener_notification branch is existing code and the one inside !listener_notification is new code.

Reply:
Maybe I'm missing something. I saw big code movements. Maybe you forgot to rebase or push?

Reply:
The move was splitting common code from apply_state_locally to do_apply_state_locally. I squashed it in new version: https://github.com/scylladb/scylla/pull/6848#issuecomment-663314943. Hope it is easier to review now.

Reply:
It would be easier if you first create do_apply_state_locally() without the listener_notification flag, which is a pure code movement. Then the this patch adds listener_notification and it stands out from the code movement in the previous patch.

Reply:
OK. I added a version do_apply_state_locally without listener_notification.